pymongo/pool.py

```diff
@@ -624,47 +667,48 @@ class Connection:
       - `id`: the id of this socket in it's pool
     """
 
-    def __init__(self, conn, pool, address, id):
+    def __init__(
+        self, conn: Union[socket.socket, _sslConn], pool: Pool, address: Tuple[str, int], id: int
+    ):
         self.pool_ref = weakref.ref(pool)
         self.conn = conn
         self.address = address
         self.id = id
-        self.authed = set()
         self.closed = False
         self.last_checkin_time = time.monotonic()
         self.performed_handshake = False
-        self.is_writable = False
+        self.is_writable: Optional[bool] = False
         self.max_wire_version = MAX_WIRE_VERSION
         self.max_bson_size = MAX_BSON_SIZE
         self.max_message_size = MAX_MESSAGE_SIZE
         self.max_write_batch_size = MAX_WRITE_BATCH_SIZE
         self.supports_sessions = False
-        self.hello_ok = None
+        self.hello_ok: Optional[bool] = None
         self.is_mongos = False
         self.op_msg_enabled = False
         self.listeners = pool.opts._event_listeners
         self.enabled_for_cmap = pool.enabled_for_cmap
         self.compression_settings = pool.opts._compression_settings
-        self.compression_context = None
-        self.socket_checker = SocketChecker()
-        self.oidc_token_gen_id = None
+        self.compression_context: Union[SnappyContext, ZlibContext, ZstdContext, None] = None
+        self.socket_checker: SocketChecker = SocketChecker()
```

Comment:
Is `SocketChecker` necessary here? It should be redundant here right?

Reply:
So this one is interesting. I agree that it feels redundant but without it mypy fails. I can take it away so you can see what the test failure would look like if that'd help.

Reply:
Yes that would be helpful.

Reply:
just did this commit.

Reply:
So the error this? What does SocketChecker have to do with this?:
```
Found 2 errors in 1 file (checked 85 source files)
pymongo/pool.py: note: In function "_configured_socket":
pymongo/pool.py:12[42](https://github.com/mongodb/mongo-python-driver/actions/runs/5757722911/job/15609185080#step:5:43): error: Incompatible types in assignment (expression has
type "_sslConn", variable has type "socket")  [assignment]
                    sock = ssl_context.wrap_socket(sock, server_hostname=h...
                           ^
pymongo/pool.py:12[44](https://github.com/mongodb/mongo-python-driver/actions/runs/5757722911/job/15609185080#step:5:45): error: Incompatible types in assignment (expression has
type "_sslConn", variable has type "socket")  [assignment]
                    sock = ssl_context.wrap_socket(sock)
                           ^
typecheck-mypy: exit 1 (14.39 seconds) /home/runner/work/mongo-python-driver/mongo-python-driver> mypy --install-types --non-interactive bson gridfs tools pymongo pid=1865
```

Reply:
wait interesting, the error i get locally is this:
```
typecheck-mypy: commands[1]> mypy --install-types --non-interactive --disable-error-code var-annotated --disable-error-code attr-defined --disable-error-code union-attr --disable-error-code assignment --disable-error-code no-redef --disable-error-code index --allow-redefinition --allow-untyped-globals --exclude 'test/mypy_fails/*.*' --exclude test/conftest.py test
Found 1 error in 1 file (checked 125 source files)
pymongo/pool.py:83: note: In module imported here,
test/pymongo_mocks.py:26: note: ... from here:
pymongo/network.py: note: In function "wait_for_read":
pymongo/network.py:303: error: Cannot determine type of "socket_checker"  [has-type]
                    readable = conn.socket_checker.select(sock, read=True, timeout=timeout)
                               ^~~~~~~~~~~~~~~~~~~
typecheck-mypy: exit 1 (10.22 seconds) /Users/iris.ho/GitHub/mongo-python-driver> mypy --install-types --non-interactive --disable-error-code var-annotated --disable-error-code attr-defined --disable-error-code union-attr --disable-error-code assignment --disable-error-code no-redef --disable-error-code index --allow-redefinition --allow-untyped-globals --exclude 'test/mypy_fails/*.*' --exclude test/conftest.py test pid=95467
.pkg: _exit> python /Users/iris.ho/Library/Python/3.9/lib/python/site-packages/pyproject_api/_backend.py True setuptools.build_meta
  typecheck-mypy: FAIL code 1 (18.80=setup[6.96]+cmd[1.62,10.22] seconds)
  evaluation failed :( (18.87 seconds)
  ```

Reply:
Hmm when I run tox locally I get a different error:
```
pymongo/pool.py:83: note: In module imported here,
test/pymongo_mocks.py:26: note: ... from here:
pymongo/network.py: note: In function "wait_for_read":
pymongo/network.py:303: error: Cannot determine type of "socket_checker"  [has-type]
                    readable = conn.socket_checker.select(sock, read=True, timeout=timeout)
                               ^~~~~~~~~~~~~~~~~~~
Found 1 error in 1 file (checked 125 source files)
```

Reply:
Yeah it does look like adding SocketChecker as a type here is required.